File: android/src/main/java/app/independo/capacitorvoicerecorder/VoiceRecorder.java
package app.independo.capacitorvoicerecorder;

import android.Manifest;
import app.independo.capacitorvoicerecorder.adapters.PermissionChecker;
import app.independo.capacitorvoicerecorder.adapters.RecordDataMapper;
import app.independo.capacitorvoicerecorder.adapters.RecorderPlatform;
import app.independo.capacitorvoicerecorder.core.ErrorCodes;
import app.independo.capacitorvoicerecorder.core.Messages;
import app.independo.capacitorvoicerecorder.core.RecordData;
import app.independo.capacitorvoicerecorder.core.RecordOptions;
import app.independo.capacitorvoicerecorder.core.ResponseFormat;
import app.independo.capacitorvoicerecorder.core.ResponseGenerator;
import app.independo.capacitorvoicerecorder.platform.DefaultRecorderPlatform;
import app.independo.capacitorvoicerecorder.service.VoiceRecorderService;
import app.independo.capacitorvoicerecorder.service.VoiceRecorderServiceException;
import com.getcapacitor.PermissionState;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;

@CapacitorPlugin(
    name = "VoiceRecorder",
    permissions = { @Permission(alias = VoiceRecorder.RECORD_AUDIO_ALIAS, strings = { Manifest.permission.RECORD_AUDIO }) }
)
/** Capacitor bridge for the VoiceRecorder plugin. */
public class VoiceRecorder extends Plugin {

    /** Permission alias used by the Capacitor permission API. */
    static final String RECORD_AUDIO_ALIAS = "voice recording";
    /** Service layer that owns recording flows and validation. */
    private VoiceRecorderService service;
    /** Response format derived from plugin configuration. */
    private ResponseFormat responseFormat;

    @Override
    public void load() {
        super.load();
        responseFormat = ResponseFormat.fromConfig(getConfig());
        RecorderPlatform platform = new DefaultRecorderPlatform(getContext());
        PermissionChecker permissionChecker = this::doesUserGaveAudioRecordingPermission;
        service = new VoiceRecorderService(platform, permissionChecker);
    }

    /** Checks whether the device can record audio. */
    @PluginMethod
    public void canDeviceVoiceRecord(PluginCall call) {
        call.resolve(ResponseGenerator.fromBoolean(service.canDeviceVoiceRecord()));
    }

    /** Requests microphone permission or returns success if already granted. */
    @PluginMethod
    public void requestAudioRecordingPermission(PluginCall call) {
        if (service.hasAudioRecordingPermission()) {
            call.resolve(ResponseGenerator.successResponse());
        } else {
            requestPermissionForAlias(RECORD_AUDIO_ALIAS, call, "recordAudioPermissionCallback");
        }
    }

    /** Forwards permission results back to the JS call. */
    @PermissionCallback
    private void recordAudioPermissionCallback(PluginCall call) {
        this.hasAudioRecordingPermission(call);
    }

    /** Returns whether the app has microphone permission. */
    @PluginMethod
    public void hasAudioRecordingPermission(PluginCall call) {
        call.resolve(ResponseGenerator.fromBoolean(service.hasAudioRecordingPermission()));
    }

    /** Starts a recording session. */
    @PluginMethod
    public void startRecording(PluginCall call) {
        try {
            String directory = call.getString("directory");
            String subDirectory = call.getString("subDirectory");
            RecordOptions options = new RecordOptions(directory, subDirectory);
            service.startRecording(
                options,
                () -> notifyListeners("voiceRecordingInterrupted", null),
                () -> notifyListeners("voiceRecordingInterruptionEnded", null)
            );
            call.resolve(ResponseGenerator.successResponse());
        } catch (VoiceRecorderServiceException exp) {
            call.reject(toLegacyMessage(exp.getCode()), exp.getCode(), exp);
        }
    }

    /** Stops recording and returns the recording payload. */
    @PluginMethod
    public void stopRecording(PluginCall call) {
        try {
            RecordData recordData = service.stopRecording();
            if (responseFormat == ResponseFormat.NORMALIZED) {
                call.resolve(ResponseGenerator.dataResponse(RecordDataMapper.toNormalizedJSObject(recordData)));
            } else {
                call.resolve(ResponseGenerator.dataResponse(RecordDataMapper.toLegacyJSObject(recordData)));
            }
        } catch (VoiceRecorderServiceException exp) {
            call.reject(toLegacyMessage(exp.getCode()), exp.getCode(), exp);
        }
    }

    /** Pauses an active recording session if supported. */
    @PluginMethod
    public void pauseRecording(PluginCall call) {
        try {
            call.resolve(ResponseGenerator.fromBoolean(service.pauseRecording()));
        } catch (VoiceRecorderServiceException exception) {
            call.reject(toLegacyMessage(exception.getCode()), exception.getCode(), exception);
        }
    }

    /** Resumes a paused recording session if supported. */
    @PluginMethod
    public void resumeRecording(PluginCall call) {
        try {
            call.resolve(ResponseGenerator.fromBoolean(service.resumeRecording()));
        } catch (VoiceRecorderServiceException exception) {
            call.reject(toLegacyMessage(exception.getCode()), exception.getCode(), exception);
        }
    }

    /** Returns the current recording status. */
    @PluginMethod
    public void getCurrentStatus(PluginCall call) {
        call.resolve(ResponseGenerator.statusResponse(service.getCurrentStatus()));
    }

    /** Checks whether the app has the RECORD_AUDIO permission. */
    private boolean doesUserGaveAudioRecordingPermission() {
        return getPermissionState(VoiceRecorder.RECORD_AUDIO_ALIAS).equals(PermissionState.GRANTED);
    }

    /** Maps canonical error codes back to legacy error messages. */
    private String toLegacyMessage(String canonicalCode) {
        if (ErrorCodes.DEVICE_CANNOT_VOICE_RECORD.equals(canonicalCode)) {
            return Messages.CANNOT_RECORD_ON_THIS_PHONE;
        }
        return canonicalCode;
    }
}

File: ios/Sources/VoiceRecorder/Bridge/VoiceRecorder.swift
import Foundation
import AVFoundation
import Capacitor

typealias PermissionRequester = (@escaping (Bool) -> Void) -> Void
typealias PermissionStatusProvider = () -> AVAudioSession.RecordPermission

/// Capacitor bridge for the VoiceRecorder plugin.
@objc(VoiceRecorder)
public class VoiceRecorder: CAPPlugin, CAPBridgedPlugin {
    /// Plugin identifier used by Capacitor.
    public let identifier = "VoiceRecorder"
    /// JavaScript name used for the plugin proxy.
    public let jsName = "VoiceRecorder"
    /// Supported plugin methods exposed to the JS layer.
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "canDeviceVoiceRecord", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestAudioRecordingPermission", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "hasAudioRecordingPermission", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startRecording", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopRecording", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "pauseRecording", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "resumeRecording", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getCurrentStatus", returnType: CAPPluginReturnPromise),
    ]

    /// Service layer that performs recording operations.
    private var service: VoiceRecorderService?
    /// Response format derived from plugin configuration.
    private var responseFormat: ResponseFormat = .legacy
    /// Permission requester used by the bridge.
    private var permissionRequester: PermissionRequester = { completion in
        AVAudioSession.sharedInstance().requestRecordPermission(completion)
    }
    /// Permission status provider used by the bridge.
    private var permissionStatusProvider: PermissionStatusProvider = {
        AVAudioSession.sharedInstance().recordPermission
    }

    /// Initializes dependencies after the plugin loads.
    public override func load() {
        super.load()
        responseFormat = ResponseFormat(config: getConfig())
        service = VoiceRecorderService(
            platform: DefaultRecorderPlatform(),
            permissionChecker: { [weak self] in
                self?.doesUserGaveAudioRecordingPermission() ?? false
            }
        )
    }

    func configureForTesting(
        service: VoiceRecorderService?,
        responseFormat: ResponseFormat? = nil,
        permissionRequester: PermissionRequester? = nil,
        permissionStatusProvider: PermissionStatusProvider? = nil
    ) {
        self.service = service
        if let responseFormat = responseFormat {
            self.responseFormat = responseFormat
        }
        if let permissionRequester = permissionRequester {
            self.permissionRequester = permissionRequester
        }
        if let permissionStatusProvider = permissionStatusProvider {
            self.permissionStatusProvider = permissionStatusProvider
        }
    }

    /// Returns whether the device can record audio.
    @objc func canDeviceVoiceRecord(_ call: CAPPluginCall) {
        let canRecord = service?.canDeviceVoiceRecord() ?? false
        call.resolve(ResponseGenerator.fromBoolean(canRecord))
    }

    /// Requests microphone permission from the user.
    @objc func requestAudioRecordingPermission(_ call: CAPPluginCall) {
        permissionRequester { granted in
            if granted {
                call.resolve(ResponseGenerator.successResponse())
            } else {
                call.resolve(ResponseGenerator.failResponse())
            }
        }
    }

    /// Returns whether the app has microphone permission.
    @objc func hasAudioRecordingPermission(_ call: CAPPluginCall) {
        let hasPermission = service?.hasAudioRecordingPermission() ?? false
        call.resolve(ResponseGenerator.fromBoolean(hasPermission))
    }

    /// Starts a recording session with optional file output.
    @objc func startRecording(_ call: CAPPluginCall) {
        guard let service = service else {
            call.reject(Messages.FAILED_TO_RECORD, ErrorCodes.failedToRecord)
            return
        }

        let directory: String? = call.getString("directory")
        let subDirectory: String? = call.getString("subDirectory")
        let recordOptions = RecordOptions(directory: directory, subDirectory: subDirectory)
        do {
            try service.startRecording(
                options: recordOptions,
                onInterruptionBegan: { [weak self] in
                    self?.notifyListeners("voiceRecordingInterrupted", data: [:])
                },
                onInterruptionEnded: { [weak self] in
                    self?.notifyListeners("voiceRecordingInterruptionEnded", data: [:])
                },
                onVolumeChanged: { [weak self] volume in
                    // volume is a Float between 0.0 and 1.0
                    self?.notifyListeners("volumeChanged", data: ["volume": volume])
                }
            )
            call.resolve(ResponseGenerator.successResponse())
        } catch let error as VoiceRecorderServiceError {
            call.reject(toLegacyMessage(error.code), error.code, error.underlyingError ?? error)
        } catch {
            call.reject(Messages.FAILED_TO_RECORD, ErrorCodes.failedToRecord, error)
        }
    }

    /// Stops recording and returns the audio payload.
    @objc func stopRecording(_ call: CAPPluginCall) {
        guard let service = service else {
            call.reject(Messages.FAILED_TO_FETCH_RECORDING, ErrorCodes.failedToFetchRecording)
            return
        }

        service.stopRecording { [weak self] result in
            DispatchQueue.main.async {
                guard let self = self else {
                    call.reject(Messages.FAILED_TO_FETCH_RECORDING, ErrorCodes.failedToFetchRecording)
                    return
                }

                switch result {
                case .success(let recordData):
                    let payload: Dictionary<String, Any>
                    if self.responseFormat == .normalized {
                        payload = RecordDataMapper.toNormalizedDictionary(recordData)
                    } else {
                        payload = RecordDataMapper.toLegacyDictionary(recordData)
                    }
                    call.resolve(ResponseGenerator.dataResponse(payload))
                case .failure(let error):
                    call.reject(self.toLegacyMessage(error.code), error.code, error.underlyingError ?? error)
                }
            }
        }
    }

    /// Pauses a recording session if supported.
    @objc func pauseRecording(_ call: CAPPluginCall) {
        guard let service = service else {
            call.reject(Messages.RECORDING_HAS_NOT_STARTED, ErrorCodes.recordingHasNotStarted)
            return
        }

        do {
            call.resolve(ResponseGenerator.fromBoolean(try service.pauseRecording()))
        } catch let error as VoiceRecorderServiceError {
            call.reject(toLegacyMessage(error.code), error.code, error.underlyingError ?? error)
        } catch {
            call.reject(Messages.FAILED_TO_RECORD, ErrorCodes.failedToRecord, error)
        }
    }

    /// Resumes a paused recording session if supported.
    @objc func resumeRecording(_ call: CAPPluginCall) {
        guard let service = service else {
            call.reject(Messages.RECORDING_HAS_NOT_STARTED, ErrorCodes.recordingHasNotStarted)
            return
        }

        do {
            call.resolve(ResponseGenerator.fromBoolean(try service.resumeRecording()))
        } catch let error as VoiceRecorderServiceError {
            call.reject(toLegacyMessage(error.code), error.code, error.underlyingError ?? error)
        } catch {
            call.reject(Messages.FAILED_TO_RECORD, ErrorCodes.failedToRecord, error)
        }
    }

    /// Returns the current recording status.
    @objc func getCurrentStatus(_ call: CAPPluginCall) {
        let status = service?.getCurrentStatus() ?? .NONE
        call.resolve(ResponseGenerator.statusResponse(status))
    }

    /// Returns whether AVAudioSession reports granted permission.
    func doesUserGaveAudioRecordingPermission() -> Bool {
        return permissionStatusProvider() == AVAudioSession.RecordPermission.granted
    }

    /// Maps canonical error codes back to legacy messages.
    private func toLegacyMessage(_ canonicalCode: String) -> String {
        if canonicalCode == ErrorCodes.deviceCannotVoiceRecord {
            return Messages.CANNOT_RECORD_ON_THIS_PHONE
        }
        return canonicalCode
    }
}
